File: union_find.js
'use strict';

var fs          = require('fs');
var path        = require('path');
const fileName  = './uf_data.txt';

class UF{
  constructor(N){
    this.count = N;
    let arr = new Array();
    for (let i=0; i<N; i++){
      arr[i] = i;
    }
    this.ids = arr;
  }
  _count(){
    return this.count;
  }
  _connected(p, q){
    return this.find(p) == this.find(q);
  }
  _find(int) {
    // fill in
  }
  _union(p, q){
    // fill in
  }
};

let file    = fs.readFileSync(fileName, 'utf8');
let lines   = file.split('\n');
let N       = lines.length-1;
let uf      = new UF(N);
var count   = uf._count();
console.log('UF', uf, count);

lines.forEach((line) => {
  if (! line) {return;}
  let parts = line.split(' ');
  console.log('PARTS', parts);
});

File: union_find.rb
class UnionFindQuickUnion

	attr_accessor :id

	def initialize(n)
		@id = *(0..n)
		@sz = *(0..n)
	end

	def connected?(p, q)
		return @id[p] == @id[q]
	end

	def root(i)
		while i != @id[i]
			@id[i] = @id[@id[i]] #path compression
			i = @id[i]
		end
		return i
	end

	def union(p, q)
		i = root(p)
		j = root(q)
		# @id[i] = j <- non-weighted
		return if i == j# weighted QU
		if @sz[i] < @sz[j]
			@id[i] = j
			@sz[j] += @sz[i]
		else
			@id[j] = i
			@sz[i] += @sz[j]
		end 
	end
end

def assemble_data
	f = File.open('uf_data.txt')
	f.readlines[0..-2]
end

def sanitize(a)
	a.map!{|p| p.split(' ').map!{|n| n.to_i}}
end

data = sanitize(assemble_data)
uf = UnionFindQuickUnion.new(9)
data.each {|d| uf.union(d[0],d[1])}
